---
title: "PulmonDB - COVID19. Daniel Blanco Data Processing"
author: "Monica Padilla"
date: "January, 2021"
output:
  html_document
---


## **Data Processing** {.tabset}

Data generated by Daniel Blanco et al[1]


\

### **Make fastqc's**

```{r eval=FALSE}
ssh -Y mpadilla@dna.liigh.unam.mx
umask 2 # para dar permisos
```

Make directories :

```{r eval=FALSE}
cd /mnt/Citosina/amedina/mpadilla/
mkdir COVID19
mkdir COVID19/Data
cd COVID19/Data/
ln -s /mnt/Citosina/amedina/aaltamirano/Data/COVID19_SRP253951/fastq/ fastq
mkdir fastqc
```

fastqc files were made with the following sge `make_fastqc.sge` :

```{r eval=FALSE}
ls fastq/ > fastqc/makefastqc
perl -pe 's/^/fastqc ..\/fastq\//' fastqc/makefastqc
```

```
#!/bin/bash
# Use current working directory
#$ -cwd
#
# Join stdout and stderr
#$ -j y
#
# Run job through bash shell
#$ -S /bin/bash
#
#You can edit the scriptsince this line
#
# Your job name
#$ -N C19DB_fastqc
#
# Send an email after the job has finished
#$ -m e
#$ -M monicapadilla905@gmail.com
#
# If modules are needed, source modules environment (Do not delete the next line):
. /etc/profile.d/modules.sh
#
# Add any modules you might require:
module load fastqc/0.11.3
#
# Write your commands in the next line
fastqc ../fastq/SRR11412215_1.fastq
fastqc ../fastq/SRR11412216_1.fastq
fastqc ../fastq/SRR11412217_1.fastq
fastqc ../fastq/SRR11412218_1.fastq
fastqc ../fastq/SRR11412219_1.fastq
...
```

```{r eval=FALSE}
qsub make_fastqc.sge
mv fastq/*fastqc* fastqc/
```

\

***

### **Group runs by experiment**

Karen got the SRA metadata `mnt/Citosina/amedina/mpadilla/COVID19/Data/metadata/SraRunTable_datosDaniel.txt` from [here](https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA615032&o=acc_s%3Aa)

\

cd = `/mnt/Citosina/amedina/mpadilla/COVID19/Data`

Get columns of run and experiments, edit experiments names to shorten them, create that file:

```{r eval=FALSE}
mkdir metadata/inter-files/
cut -d, -f1,28 metadata/SraRunTable_datosDaniel.txt | perl -pe 's/ /_/g' | sed '1d' | perl -pe 's/_with_a_/_w_/' | perl -pe 's/_cells//' | perl -pe 's/_after//' | perl -pe 's/\((.*)\)/$1/' | perl -pe 's/human_ACE2/hACE2/' | perl -pe 's/_trasnduced//' | perl -pe 's/,/\t/' > metadata/inter-files/runs_experiments.tsv
```

* perl was used to change the names of experiments by substituting characters
* sed was used to take off the header

Get columns of runs, samples and experiments, edit experiments names to shorten them as before, create that file:

```{r eval=FALSE}
cut -d, -f1,27,28 metadata/SraRunTable_datosDaniel.txt | perl -pe 's/ /_/g' | sed '1d' | perl -pe 's/_with_a_/_w_/' | perl -pe 's/_cells//' | perl -pe 's/_after//' | perl -pe 's/\((.*)\)/$1/' | perl -pe 's/human_ACE2/hACE2/' | perl -pe 's/_trasnduced//' | perl -pe 's/,/\t/g' | sort -k 3 > metadata/inter-files/runs_samples_experiments.tsv
```

* `sort -k 3` sorts lines according to the 3rd column

\

Generate directories of experiments and move their fastq runs there (based on `runs_experiments.tsv` ):

```{r eval=FALSE}
cat metadata/inter-files/runs_experiments.tsv | perl -pe 'm/(.*)\t(.*)/; system("mkdir","fastq/".$2) unless (-e "fastq/".$2); system("mv","fastq/".$1."_1.fastq","fastq/".$2."/")'
```

grouped data:

```
[mpadilla@chromatin Data]$ ls fastq
HPIV3_infected_A549                           Mock_treated_A549_w_vector_expressing_hACE2
human_IFNB_treated_NHBE                       Mock_treated_Calu-3
IAVdNS1_infected_NHBE                         Mock_treated_NHBE
IAV_infected_4_month_old_Ferret               RSV_infected_A549
IAV_infected_A549                             SARS-CoV-2_infected_4_month_old_Ferret
IAV_infected_NHBE                             SARS-CoV-2_infected_A549
Lung_biopsy_for_heatly_negative_control       SARS-CoV-2_infected_A549_w_vector_expressing_hACE2
Lung_sample_from_postmortem_COVID-19_patient  SARS-CoV-2_infected_A549_w_vector_expressing_hACE2_1hr_Ruxolitinib_pre-treatment
Mock_treated_4_month_old_Ferret               SARS-CoV-2_infected_Calu-3
Mock_treated_A549                             SARS-CoV-2_infected_NHBE

```

do the same for fastqc's

```{r eval=FALSE}
cat metadata/inter-files/runs_experiments.tsv | perl -pe 'm/(.*)\t(.*)/; system("mkdir","fastqc/".$2) unless (-e "fastqc/".$2); system("mv","fastqc/".$1."_1_fastqc.html","fastqc/".$2."/")'
```

\

***

### **Quality Check: fastqc**

**By experiment:**

#### **HPIV3_infected_A549**

.                      | SRR11517756 | SRR11517757 | SRR11517758
-----------------------|-----------|-----------|------------
per base seq quality   | good, score drops below 28 at 135-139  | good, score drops below 28 at 135-139 | good, as before
per seq quality scores | good, peak at ~35 score | good, peak at ~35 score | good, as before
per base seq content   | good, lines are messy 1-15, but may be due to library prep bias, lines are parallel 15-150 pos | good, same as SRR11517756 | good, same as SRR11517756
per seq GC content     | warning-okay, GC count distrib is somewhat within the Theoretical but has peaks, but there may be biases in transcripted reads | warning-okay, GC count distrib coincides with theoretical | okay, similar to SRR11517756
per base N content     | good, no N content | good, no N content | good, as before
seq length distribution| warning, peak of length at 150-152 but another small distrib from 80-140 | okay, peak at 150-152 | warning, similar to SRR11517756
seq dupl levels        | good, duplication peaks at >10->1k levels | good, duplication peaks at >10->1k levels | good, as before
overrepresented seqs   | okay, there are overrepresented seqs | okay, one overrep seq | okay, a NN.. overrep seq adn two more
adapter content        | good, no adapter content | good, no adapter content | good, no adapter content
kmer content           | okay, there are peaks of overrepresented seqs mostly at the end 139 bp | okay, most overrep kmers from 1-5 pos | okay, peaks at 7-15 bp and at 135-139bp

[per base seq content RNA-seq bias](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/4%20Per%20Base%20Sequence%20Content.html)

Note: as before is as the sample run before

\

#### **human_IFNB_treated_NHBE**


.                      | SSRR11517807 | SRR11517808 | SRR11517809 | SRR11517810 | SRR11517811 | SRR11517812 | SRR11517813 | SRR11517814
-----------------------|-----------|-----------|------------|--|--|--|--|--|--
per base seq quality   |  good, mean quality drops below 30 at 135bp  | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before
per seq quality scores | good, peak at 34 score | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before
per base seq content   | okay, lines run parallel 15-150, 1-15 lines are messy | okay, as before |  okay, as before | okay, as before | okay, messy 1-15, parallel from then but G has an increment | okay, as SRR11517811 | okay, as SRR11517811 | okay, as SRR11517811 
per seq GC content     | warning-okay, generally GC count falls within theoretical distrib | w-okay, as before | w-okay, as before | w-okay, as before | bad, GC count looks uniform but has a peak at 63% and 0% | bad, as SRR11517811 | bad, as SRR11517811 | bad, as SRR11517811 
per base N content     | good, virtually no N content | good, as before | good, as before | good, as before | good, a bit N content at the start | good, as before | good, as before | good, as before
seq length distribution| okay, length peak at 150-152 | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before
seq dupl levels        | okay, only a peak at >10 dup level | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before
overrepresented seqs   | bad, only two overrep seqs: NN.. and adapter | bad, as before | bad, as before | bad, as before | bad-okay, 3 overrep seqs: adapter, NN.., GG.. | b-okay, as SRR11517811 | b-okay, 4 overrep seqs: 2 adapters, NN.., G.. | bad-okay, as SRR11517811
adapter content        | good, virtually no adapter content | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before
kmer content           | bad-okay, a lot of kmers 1-15bp | b-okay, as before | b-okay, as before | b-okay, as before | b-okay, as before but more kmers | b-okay, as SRR11517811 | b-okay, as SRR11517811 | b-okay, as SRR11517811



.                      | SRR11517815 | SRR11517816 | SRR11517817 | SRR11517818 | SRR11517819 | SRR11517820 | SRR11517821 | SRR11517822
-----------------------|-----------|-----------|-------------|--|--|--|--|--|--
per base seq quality   | good, mean quality drops below 30 at 135-139bp | good, mean drops below 30 at 150 | good, mean drops below 30 at 135-139 | good, as before | good, as before | good, as before | good, as before | good, as before
per seq quality scores | good, peak at 34 score | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before
per base seq content   | okay, messy lines 1-15bp possibly due to RNA-seq bias, parallel or crossing lines 15-150bp | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before
per seq GC content     | bad-okay, GC count seems normal but peak at 63% and 0% | b-okay, as before | b-okay, as before | b-okay, as before | b-okay, as before | b-okay, as before | b-okay, as before | b-okay, as before
per base N content     | good, virtually no N content | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before
seq length distribution| okay, peak at 150-152bp and bit at 30-40 | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before
seq dupl levels        | warning-okay, only one peak at >10 dup level, perhaps very few duplicated | w-okay, as before | w-okay, as before | w-okay, as before | w-okay, as before | w-okay, as before | w-okay, as before | w-okay, as before
overrepresented seqs   | bad-okay, 3 overrep seqs: adapter, NN.., GG.. | b-okay, as before | b-okay, as before | b-okay, as before | okay, as before | bad-okay, 3 overrep seqs: adapter, NN.. | b-okay, as SRR11517820 | b-okay, as before
adapter content        | good, no adapter content | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before
kmer content           | bad-okay, kmer peaks at 1-15bp at >80 | b-okay, as before | b-okay, as before | b-okay, as before | b-okay, as before but >100 | b-okay, as SRR11517819 | b-okay, as before |  b-okay, as before



.                      | SRR11517823 | SRR11517824 | SRR11517825 | SRR11517826 | SRR11517827 | SRR11517828 | SRR11517829 | SRR11517830
-----------------------|-----------|-----------|------------|--|--|--|--|--|--
per base seq quality   | good, mean quality drops below 30 at 150bp | good, as before | good, mean drops below 30 at ~130 | good, mean drops below 30 at 135-139 | good, as before | good, as before | good, mean drops below 30 at 150 | good, as before 
per seq quality scores | good, peak at 34 score | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before 
per base seq content   | okay, messy lines 1-15, crossing or parallel lines 15-150 | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before 
per seq GC content     | bad-okay, GC counts seem normal but small peak 63& and a smaller at 0% | b-okay, as before | b-okay, as before | b-okay, as before | warning-okay, GC counts coincides with theoretical distrib but small peaks at 63% and 0& | w-okay, as SRR11517827 | w-okay, as before | w-okay, as before 
per base N content     | good, no N content | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before 
seq length distribution| okay, length peak at 150-152bp | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before | okay, as before 
seq dupl levels        | warning-okay, only one peak at >10 dupl level | w-okay, as before | w-okay, as before | w-okay, as before | w-okay, as before | w-okay, as before |  w-okay, as before | w-okay, as before 
overrepresented seqs   | bad, 2 overrep seqs: adapter and NN.. | bad, as before | bad, as before | bad, as before | warning-okay, as before | w-okay, 3 overrep seqs: adapter, NN.., GG.. | w-okay, as SRR11517828 | w-okay, as before 
adapter content        | good, no adapter content | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before | good, as before 
kmer content           | bad-okay, kmers peaks 1-15 at >80 | b-okay, as before | b-okay, as before | b-okay, as before | b-okay as before but >60 | b-okay, as SRR11517827 | b-okay, as before | b-okay, as before 

Conclusion points: 

* runs are very similar to one another
* sequences will be trimmed at 30 phred score, leaving most seqs within 130-150 length
* seq composition seems good except at the start but this may be a bias due to library RNA-seq preparation
* GC content is overall normal but has peaks at 63% and 0%, however this might not be a problem, there might be a bias in the cell for certain transcripts
* good scores, almost no N content, length is okay
* very few duplicates, might be interesting at enrichment analysis

\

#### **IAVdNS1_infected_NHBE**

.                      | Notes SRR11517791-798 | SRR11517799-802 | SRR11517803-806
-----------------------|----------------------|------------------|-------------
per base seq quality   | good, mean quality drops below 30 at 135-139 or 150bp but many reads drop from ~60bp | good mean drops at 150bp | -
per seq quality scores | good, broad peak score at 34 | good, peak at 34 | -
per base seq content   | okay, messy lines 1-15 but may be due to RNA-seq bias, crossing or parallel lines 15-150 | okay, messy at the beginning but then parallel lines, G rises by the middle | bad, very messy 1-15, less messy 15-60 and G increment by ten and more from then
per seq GC content     | bad-okay, GC count seems normal or coincides with theoretical but with peaks, specially one at ~63% (in a few) and a smaller one at 0% | strange normal, big peak at 63% and 0% | almost normal but much more accentuated peak at 63% and 0%
per base N content     | good, virtually no N content | good, no N | -
seq length distribution| okay, peak at 150-152 but some within 100-150 | wider 2nd distribution within 80-140 (1st one is a peak at 150-152) | -
seq dupl levels        | okay, peak at >10 dupl level, very few I think | as before + smaller peak at >100 | -
overrepresented seqs   | bad, overrep seqs: adapter(1-3), NNN..., GGG... (0-2) | as before | as before
adapter content        | good, no adapter content | good | -
kmer content           | bad-okay, kmer peaks 1-15 | as before | bigger kmer peaks

Conlusion points:

* sequences will be trimmed at 30 and will be 135-150 in length except some that are very small
* virtually no adapter, nor N content, good quality base call peak, few duplicated
* strange GC counts distibutions, almost normal in many but persistent peak at 63%
* per base content got worse in these reads, but overall the same trend: messy lines at the start and parallel from there except some

\

#### **IAV_infected_4_month_old_Ferret**

.                      | SRR11517719-724 | SRR11517879-886
-----------------------|-----------|------------------
per base seq quality   | good, mean quality drops at 150bp | -
per seq quality scores | good, score peak at >34 | -
per base seq content   | okay, messy lines 1-15, nicely parallel 15-150 | okay, messy 1-15, crossing but parallel 15-150
per seq GC content     | okay, GC counts coincided with theoretical but with small left bias | okay, GC counts distrib almost normal but many peaks
per base N content     | good, no N content | -
seq length distribution| warning, peak at 150-152 and 2nd small distrib within 90-140 | warning, peak at 150-152 and some reads within 120-140 
seq dupl levels        | warning-okay, few duplicates at >10 level | okay, many duplicates at >100, >1k, >10k levels
overrepresented seqs   | good-okay, no overrep seqs or one | okay, a lot of overrep seqs, may be transcrips
adapter content        | good, no adapter | -
kmer content           | okay, few (~15) kmer peaks at 1-5 and some (~30) at 139bp | okay, many kmer peaks at 1-15

Conlusion points:

* better quality than previous experiments
* more duplicates and overrep seqs in some of them 
* same trend in per base content

\

#### **IAV_infected_A549**

.                      | SRR11412287-290 | SRR11412291-294
-----------------------|-----------|----------
per base seq quality   | very good, virtually >= 30 score all along | good, mean quality >=30
per seq quality scores | good, peak at >34 | -
per base seq content   | okay, 1-15 messy lines, parallel or crossing 15-150 | bad, messy 1-15, crossing 15-75, G rises a lot 
per seq GC content     | bad-okay, GC counts distrib is somewhat normal but big peaks at 63% and 0% | okay-bad,  GC distrib has a big peak at ~63% and another at 0%
per base N content     | good, virtually no N content | warning, few N content
seq length distribution| warning-okay, peak at 150 length but some within 110-144 and few within 30-40 | -
seq dupl levels        | okay-good, few duplicates at >10, >100 and >10k | -
overrepresented seqs   | okay, some overrep seqs | -
adapter content        | good, no adapter content | -
kmer content           | okay, some kmers at the start | -

Conclusion points:

* very good base call quality
* GC content distribution is nuts
* same trend in per base content but some are messier
* many duplicates

\

#### **IAV_infected_NHBE**

.                      | SRR11517775-782 | SRR11517783-790  
-----------------------|-----------------|----------------
per base seq quality   | good, mean quality at 28 until 6pb, then rises and drops below 30 at 150 | good, mean quality >=30
per seq quality scores | good, broad peak at >34 score | good, peak at >34
per base seq content   | okay, messy lines 1-15, crossing 15-60 then G rises | okay, messy 1-15, parallel 15-150
per seq GC content     | bad-okay, distrib within 31-67, with peaks at ~60% and 0% | okay, GC count somewaht normal with peaks at ~60% an 0%
per base N content     | warning, N content counts at 9 until 30bpp | good, almost no N content
seq length distribution| warning, peak at 150-152 but other two smaller distribs, one at 30-44 and other 80-140 | warning, peak at 150-152 but many or some within 80-140 and some within 30-44
seq dupl levels        | okay, duplicates at >10, >100, >10k | okay, small peak at >10 dup lvl
overrepresented seqs   | bad, two overrep seqs: NN.., >=2 adapters | bad, two overrep seqs: NN.., 1 adapter
adapter content        | good, virtually no adapter content | -
kmer content           | okay, kmer peaks 1-15 |  - 

Conclusion points:

* base call quality is good
* some duplicates
* same trend in GC content peaks as before
* more N content

\

#### **Lung_biopsy_for_heatly_negative_control**

.                      | SRR11517725-728 | SRR11517729-732 
-----------------------|-----------------|-----------------
per base seq quality   | good, mean quality >=30 score | -
per seq quality scores | good, score peak >34 | -
per base seq content   | okay, messy lines 1-15, parallel 15-150 | -
per seq GC content     | okay, GC count seems normal but with peak at 63% | okay, same as before but bigger peak at 63%
per base N content     | good, no apparent N content | -
seq length distribution| okay, peak at 150-152 length but some within 80-140 | -
seq dupl levels        | okay, small duplicates peaks >10,>100,>1k | -
overrepresented seqs   | okay-bad, some overrep seqs, 2 are adpaters | bad, two adapters
adapter content        | good, virtually no adapter content | -
kmer content           | okay, kmer peaks 1-15bp | -

*-* : same as before

Conlusion points:

* good base quality
* some duplicates
* same trend in GC content (peak ~63%)

\

#### **Lung_sample_from_postmortem_COVID-19_patient**

.                      | SRR11517733-740 
-----------------------|-----------------
per base seq quality   | good, mean quality >=~30,  strangely mean drops only within 45-79bp
per seq quality scores | good, peak >=34 score
per base seq content   | very bad, lines never are parallel, just different peaks and G takes over from 75bp 
per seq GC content     | bad, only a peak within 59-62
per base N content     | good, no apparent N content
seq length distribution| good, peak at 150-152
seq dupl levels        | okay, big peak at >10k
overrepresented seqs   | very bad, many adapters 
adapter content        | good, ??
kmer content           | warning, 4 big kmer peaks at the start

Conclusion points:

* good base call quality
* super high G content, GC distrib is a peak and per base seq content is a mess
* many duplications but many overrepresented adapters

\

#### **Mock_treated_4_month_old_Ferret**

.                      | SRR11517683-686 | SRR11517687-698 | SRR11517699-702 | SRR11517831-834 | SRR11517835-838 | SRR11517847-854 | SRR11517863-870
-----------------------|-----------------|-----------|-------------|-------|-------|--|--
per base seq quality   | warning, mean quality drops below 30 at ~100bp | good, mean quality >=30 | good, mean quality drops below 30 at 150 | good, mean quality drops below 30 at 150bp | - | good, mena drops below 30 at 135-139 | very good, mean >=30
per seq quality scores | good, most seqs are >28 score | good, most seqs are >=30 | - | - | - | - | very good 
per base seq content   | okay, messy lines until 15bp, then somewhat parallel | okay, messy lines 1-15, then parallel until 150 bp | -, nicely parallel 15-150bp | okay, messy 1-15, parallel 15-150bp | - | - | -
per seq GC content     | warning-okay, GC counts somewhat normal within 51-87%, but with many peaks | okay, GC counts seems almost normal but has peaks, specially one at 63% | good, somewhat coincides with theoretical | okay, GC counts (35-75) somewhat coincides with theoretical but biased at the center | good, coincides with theoretical | okay, somewhat coincides with teorethical but with peaks | okay-good, -
per base N content     | good, no apparent N content | - | - | - | - | - | -
seq length distribution| warning, peak at 150 but many within 60-140 | okay, most are >150, but some are within 100-140 | - | okay, peak at 150-152 length | -, some fall within 100-140 | - | warning-okay, peak at 150-152 but some within 80-140
seq dupl levels        | okay, duplicates at >10,>100,>1k levels | okay, duplicates peaks at >10, >100, >1k | - , peaks at >10, >100 | okay, duplicate peaks >10, >100 levels | -, + small peak >1k | - | okay, few duplicated >10, >100, >500
overrepresented seqs   | okay, many overrep seqs | okay, 2 overep seqs: adapter, AA.. | okay, no overrep seqs or few | okay, some overrep seqs | - | okay, few or many overrep seqs | okay, 1-2 overep seqs 
adapter content        | good, no adapter | - | - | - | -  | - | -
kmer content           | okay, kmer peaks 1-5 and 135-139bp | kmer peaks at 1-15bp and 139 | okay small (<15) kmer peaks at 1-7bp and/or >139bp | okay, kmer peaks 1-15 | few kmer peaks 1-15 and 139 | - | -

Conclusion notes:

* overall god base call quality, seqeunces will be 100-150 in length after trimming, except some that are even 40bp in length
* same trend as before in per base seq content
* GC counts distribs seemed more normal in these reads, but with peaks
* some-many duplicated

\

#### **Mock_treated_A549**

.                      | SRR11412239-242 | SRR11412243-246 | SRR11412247-250, SRR11412263-270 | SRR11412279-286 | SRR11517674-676, SRR11517750-752
-----------------------|-----------------|-----|------|------|------|------
per base seq quality   | good, mean >=30 | - | - | - | -
per seq quality scores | good, most are >30 | - | - | - | -
per base seq content   | okay, messy 1-15, parallel 15-150 | - | -, but G rises from 75 | okay, messy 1-15, parallel 15-150 | -
per seq GC content     | okay, nicely normal but big peak at ~60% | good, coincides with theoretical | -. but big peak at ~60% | okay, almost normal but small peak at ~60% | good, coincides with theoretical but peak at 0%
per base N content     | good, no N content | - | - | - | -
seq length distribution| good, peak at 150-152 length | - | - | - | -
seq dupl levels        | okay, dup peaks >10, >100 | okay, peak >10, >10k | - | - | -
overrepresented seqs   | okay, 1 adapter | okay, some adapters | okay, 3 overep seqs + 2 adapters | okay, some overrep seqs | okay, one overrep
adapter content        | good, no adapter | - | - | - | -
kmer content           | okay, kmer peaks 1-15 | - | - | - | -


Conclusion points:

* overall good base call quality
* GC counts distrib looks fine, but with peaks
* some duplicates, some adapters

\

#### **Mock_treated_A549_w_vector_expressing_hACE2**

.                      | SRR11517680-682 | SRR11573892-903 
-----------------------|-----------------|-----------------
per base seq quality   | good, mean drops below 30 at 150bp | good, mean drops elow 30 at 135bp
per seq quality scores | good, most seqs are >30 | -
per base seq content   | okay, messy lines 1-15, parallel 15-150bp | -
per seq GC content     | okay-good, coincides with theoretical | -
per base N content     | good, no N content | -
seq length distribution| okay, peak at 150-152 | warning-akey, peak at 150 but many within 60-140bp
seq dupl levels        | okay, duplicates at >10, >100, >1k | okay, duplicates >10, >100
overrepresented seqs   | okay, 0-2 overrep seqs | -
adapter content        | good, no adapters | -
kmer content           | okay, some kmer peaks at 1-7bp | -, + at end

Conclusion points:

* good base quality, trimming will results in 135-150bp reads except some in even 60 bp 
* GC counts distrib coincides with theoretical
* some duplicates

\

#### **Mock_treated_Calu-3**

.                      | SRR11517744-746 
-----------------------|-----------------
per base seq quality   | good, mean drops below 30 at 150bp
per seq quality scores | good, most seqs are >30
per base seq content   | okay, messy lines 1-15, parallel 15-150bp
per seq GC content     | good, coincides with theoretical
per base N content     | good, no N content
seq length distribution| okay, peak at 150-152 length + some within 84-140
seq dupl levels        | okay, duplicates at >10, >100 (>1k)
overrepresented seqs   | okay, 0-1 overrep seqs
adapter content        | good, no adapters
kmer content           | okay, some kmer peaks at 1-7b and at 139

Conclusion points:

* good base quality, trimming will results in 150bp reads except some in even 80 bp 
* GC counts distrib coincides with theoretical
* some duplicates

\

#### **Mock_treated_NHBE**

.                      | SRR11412215-226, SRR11517759-774
-----------------------|---------------------------------
per base seq quality   | good, mean drops below 30 at 150bp
per seq quality scores | good, most seqs are >30
per base seq content   | okay, messy lines 1-15, parallel 15-150bp
per seq GC content     | okay-good, coincides (at least somewhat) with theoretical
per base N content     | good, no N content
seq length distribution| okay, peak at 150-152 length + 2nd distrib within 84-140
seq dupl levels        | okay, duplicates at >10, >100, >1k
overrepresented seqs   | okay, some overrep seqs
adapter content        | good, no adapters
kmer content           | okay, few kmer peaks at 1-5b and at 139

Conclusion points:

* good base quality, trimming will results in 150bp reads except some in even 80 bp 
* GC counts distrib coincides with theoretical
* some duplicates

\

#### **RSV_infected_A549**

.                      | SRR11412271-278 | SRR11517753-755
-----------------------|-----------------|-----------------
per base seq quality   | good, mean quality >=30 | -
per seq quality scores | good, most seqs >30 | -
per base seq content   | okay, messy lines 1-15, parallel 15-150, but G increments a little at 75 | okay, messy 1-15, parallel 15-150bp
per seq GC content     | okay, GC counts seems normal but has a big peak at ~63% | okay, GC counts somewhat similar to theoretical
per base N content     | good, no N content | -
seq length distribution| okay, peak at 150-152 length | -
seq dupl levels        | okay, duplicates >10, >100, >10k | okay, duplicates >10, >100, >1k
overrepresented seqs   | okay, few overrep seqs | okay, NN.. + some more
adapter content        | good, no adapter | -
kmer content           | okay, some kmer peaks at 1-15bp | -

Conclusion points:

* good base call quality, trimming wil result in 150 length reads
* GC counts have similar trend as before or normal
* some duplicates, no adapters

\

#### **SARS-CoV-2_infected_4_month_old_Ferret**

.                      | SRR11517691-694 | SRR11517695-698 | SRR11517703-718 | SRR11517839-862 | SRR11517871-870
-----------------------|-----------------|-------|----------|-------|---------
per base seq quality   | good, mean quality >=30 | - | - | good, mean drops below 30 at 139-150 | good, mean quality >=30
per seq quality scores | good, most seqs are >30 | - | - | - | -
per base seq content   | okay, messy lines 1-15, parallel 15-150bp, but G increments a little at 75 | okay, messy lines 1-15, parallel 15-150bp | - | - | -
per seq GC content     | okay, GC counts seem somewhat normal (31-75%) but with big peak at ~60% | GC counts weirdly distributed within 31-79% + peak at 63% | okay-good, coincides (somewhat) with theoretical | okay, GC counts somewhat normal within 27-75% but has peaks | -, or coincides with theoretical
per base N content     | good, no N content | - | - | - | -
seq length distribution| okay, peak at 150-152 length + some within 84-140 | - | -, other is 2nd distrib | okay, peak 150-152 and some within 100-140 | -, other is 2nd distrib
seq dupl levels        | okay, duplicates at >10, >100, >1k, >10k | - | okay, duplicates at >10, >100 | okay, duplicates >10, >100, (>1k) | okay, duplicate peak at >10
overrepresented seqs   | okay, 1-2 adapter | okay, 4 overrep seqs (1 adapter) | okay, 0-1 | okay, some overrep seqs (0-2 adapters) | okay, 0-1 adapter
adapter content        | good, no adapters | - | - | - | -
kmer content           | okay, few kmer peaks at 1-5b | - | okay, many kmers | okay, some or many kmers 1-15bpor at end | okay, few kmer peaks at 1-15b or at end

Conclusion points:

* good base call quality, trimming will result in 139-150 length reads, except some at even 84 length
* GC counts have a similar distrib to the theoretical or with peaks
* few duplicates, few adapters

\

#### **SARS-CoV-2_infected_A549**

.                      | SRR11412251-262 | SRR11517677-679 
-----------------------|-----------------|-----------------
per base seq quality   | good, mean quality >=30 | good, mean drops below 30 at 150bp
per seq quality scores | good, most seqs are >30 | -
per base seq content   | okay, messy lines 1-15, parallel 15-150bp, but G increments a little at 75 | okay, messy lines 1-15, parallel 15-150bp
per seq GC content     | okay-good, coincides with theoretical (35-75%) but has peak at 63% | good, coincides with theoretical (31-75)
per base N content     | good, no N content | -
seq length distribution| okay, peak at 150-152 length | -, some within 100-140
seq dupl levels        | okay, duplicates at >10, >100, >10k | okay, duplicates at >10, >100, >1k
overrepresented seqs   | okay, seqs: 1-2 adapter, 0-1 NN... | okay, 1 seq
adapter content        | good, no adapters | -
kmer content           | okay, few kmer peaks at 1-15b and at 139 | okay, few small kmer peas 1-5 and 139

Conclusion points:

* very good base call quality, trimming will result in 150 length reads, except some at 100 length
* GC counts have a similar distrib to the theoretical or with peak at 63% as before
* few duplicates, few adapters

\

#### **SARS-CoV-2_infected_A549_w_vector_expressing_hACE2**

.                      | SRR11517741-743, SRR11573904-915
-----------------------|-----------------|-----------------
per base seq quality   | good, mean drops below 30 at 150bp
per seq quality scores | good, most seqs are >30
per base seq content   | okay, messy lines 1-15, parallel 15-150bp
per seq GC content     | okay, GC counts somewhat normal (31-70%) but with peaks, specially one at ~50%
per base N content     | good, no N content
seq length distribution| okay, peak at 150-152 length + some or many within 84-140
seq dupl levels        | okay, duplicates at >10, >100, >1k, (>10k)
overrepresented seqs   | okay, many overrep seqs (0-1 adapter)
adapter content        | good, no adapters
kmer content           | okay, few kmer peaks at 1-5b and/or at 139

Conclusion points:

* overall good base quality, trimming will result in 150 length reads + several at at lowest 84 length
* some duplicates, few adapters
* GC counts distrib somewhat normal, with peaks

\

#### **SARS-CoV-2_infected_A549_w_vector_expressing_hACE2_1hr_Ruxolitinib_pre-treatment**

.                      | SRR11573916-927
-----------------------|-----------------
per base seq quality   | good, mean drops below 30 at 150bp
per seq quality scores | good, most seqs are >30
per base seq content   | okay, messy lines 1-15, parallel 15-150bp
per seq GC content     | okay, GC counts somewhat normal (31-63%) but with peaks, specially one at ~50%
per base N content     | good, no N content
seq length distribution| okay, peak at 150-152 length + some within 80-140
seq dupl levels        | okay, duplicates at >10, >100, >1k, >10k
overrepresented seqs   | okay, many overrep seqs 
adapter content        | good, no adapters
kmer content           | okay, few kmer peaks at 1-9b and/or at 139

Conclusion points:

* good base call quality, trimming will result in 150 bp length reads + some at at minimum 80bp
* some duplicates, no adapters
* GC counts distrib somewhat normal

\

#### **SARS-CoV-2_infected_Calu-3**

.                      | SRR11517747-749 
-----------------------|-----------------
per base seq quality   | good, mean drops below 30 at 150bp
per seq quality scores | good, most seqs are >30
per base seq content   | okay, messy lines 1-15, parallel 15-150bp
per seq GC content     | okay-good, similar to theoretical
per base N content     | good, no N content
seq length distribution| okay, peak at 150-152 length + some within 84-140
seq dupl levels        | okay, duplicates at >10, >100, >1k, >10k
overrepresented seqs   | okay, some overrep seqs 
adapter content        | good, no adapters
kmer content           | okay, many kmer peaks

Conclusion points:

* good base call quality, trimming will result in 150 bp length reads + some at at minimum 80bp
* some duplicates, no adapters
* GC counts distrib somewhat normal

\

#### **SARS-CoV-2_infected_NHBE**


.                      | SRR11412227-238
-----------------------|-----------------
per base seq quality   | good, mean drops below 30 at 150bp
per seq quality scores | good, most seqs are >30
per base seq content   | okay, messy lines 1-15, parallel 15-150bp
per seq GC content     | okay-good, coincides with theoretical (35-71%)
per base N content     | good, no N content
seq length distribution| okay, peak at 150-152 length + some within 84-140
seq dupl levels        | okay, duplicates at >10, >100, >1k
overrepresented seqs   | okay, some overrep seqs 
adapter content        | good, no adapters
kmer content           | okay, few kmer peaks at 1-9b and/or at 139

Conclusion points:

* good base call quality, trimming will result in 150 bp length reads + some at at minimum 80bp
* some duplicates, no adapters
* GC counts distrib coincides with thoretical


\

***

\

### **Trimming**

Prepare terminal, make directories:

```{r eval=FALSE}
screen -S trim
qlogin
cd /mnt/Citosina/amedina/mpadilla/
umask 2
mkdir resources/
mkdir resources/trimmomatic/
mkdir resources/trimmomatic/adapters/
```

General dir for resources: `/mnt/Citosina/amedina/mpadilla/resources/`

***

**Get adapter sequences:**

I copied the adapter sequences from [here](https://github.com/timflutre/trimmomatic/blob/master/adapters/TruSeq2-SE.fa) 
and stored them at `mpadilla/resources/trimmomatic/adapters`.

***

**Run [trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) across all samples to **

* cut adapters
* keep only high-quality base calls
* select a minimum length to avoid multi-mapping in further processing and avoid bias resulting from trimming [2], standard for rnaseq is 40-50 bp

cd: `/mnt/Citosina/amedina/mpadilla/mpadilla/COVID19/Data/fastq/`

```{r eval=FALSE}
module load trimmomatic/0.33
for file in */*.fastq.gz; do trimmomatic SE -phred33 $file ${file//_1.fastq/_1_trmd.fastq} ILLUMINACLIP:/mnt/Citosina/amedina/mpadilla/resources/trimmomatic/adapters/TruSeq-SE.fa:2:30:10 SLIDINGWINDOW:5:30 MINLEN:40; done
```

* this command runs trimmomatic on each .fastq file in `fastq/*/` 
* `SE` single stranded
* `-phred33` use of phred 33 code, detecting by seeing a # character on fastq files 
* `$file` .fastq file to trim
* `${file//_1.fastq/_1_trmd.fastq}.gz` name the trimmed fastq file should have, which is the same as the one before plus `_trmd`

The next parameters are taken on a **serial order** (i.e., the order of the factors changes the result) by the program and determine how are reads processed:

* `ILLUMINACLIP:/mnt/Citosina/amedina/mpadilla/resources/trimmomatic/adapters/TruSeq-SE.fa:2:30:10` cut adapters according to the file above
* `SLIDINGWINDOW:5:30` check in a sliding window of size 5, keep reads until average phred score in window is below 30
* `MINLEN:40` minimum length of reads

***

**Organize them in a different dir: `COVID19/Data/trimmed2/`, and group them by experiment and by samples according to `metadata/inter-files/runs_samples_experiments.tsv`:**

cd = `/mnt/Citosina/amedina/mpadilla/COVID19/Data`


```{r eval=FALSE}
mkdir trimmed2
mv fastq/*/*_trmd.fastq.gz trimmed2/
# make experiment dir, make sample dir, mv run file there
perl -pe 'm/(.*)\t(.*)\t(.*)/; system("mkdir","trimmed2/".$3) unless (-e "trimmed2/".$3); system("mkdir","trimmed2/".$3."/".$2) unless (-e "trimmed2/".$3."/".$2); system("mv","trimmed2/".$1."_1_trmd.fastq.gz","trimmed2/".$3."/".$2."/")' metadata/inter-files/runs_samples_experiments.tsv
```

\

### **Count average length of reads**

**For each sample, print the corresponding sample, the avglen and the mean sd of all reads (in all fastq.gzs) and save it in a file:**

```{r eval=FALSE}
for sample in ./GSM*; do \
for file in $sample/*.fastq; do less $file \
| perl -ne 'print if /^[ATCG]+$/' | awk '{ print length }' \
| { perl -ne 'BEGIN{$sum=0, $nreads=0, @dev=()} $nreads += 1; $sum += (split)[0]; push(@dev,(split)[0]); END{ $sumdev=0; $mean = $sum/$nreads; foreach my $val (@dev) {$sumdev+= ($val - $mean)**2}; print $mean."\t".($sumdev/$nreads)**(1/2)."\n"}' & echo $sample; }; done; done \
| perl -pe 's/(GSM.*)\n/$1\t/' | sort \
| perl -ne 'BEGIN{ $csample = "HPIV3_infected_A549/GSM4462360", $sumlen =0, $sumsd=0, $nfastqs=0 } if ((split)[0] eq $csample) { $nfastqs++; $sumlen += (split)[1]; $sumsd += (split)[2]; } else{ print $csample."\t".$sumlen/$nfastqs."\t".$sumsd/$nfastqs."\n"; $csample = (split)[0]; $nfastqs=1 ; $sumlen=(split)[1] ; $sumsd=(split)[2]} END{ print $csample."\t".$sumlen/$nfastqs."\t".$sumsd/$nfastqs."\n"}' > reads_avglen_sd_perSample.tsv
```

* `for sample in ./GSM*; do [something1]; done` : runs `[something1]` on each dir named GSM* which corresponds to a sample
* `for file in $sample/*.fastq.gz; do [something2]; done` : this was the previous `[something1]`, it runs `[something2]` for all `.fastq.gz` files in a sample dir, corresponding to the variable $sample before 

* `less $file | perl -ne 'print if /^[ATCG]+$/' | awk '{ print length }' | { perl -ne 'BEGIN{$sum=0, $nreads=0, @dev=()} $nreads += 1; $sum += (split)[0]; push(@dev,(split)[0]); END{ $sumdev=0; $mean = $sum/$nreads; foreach my $val (@dev) {$sumdev+= ($val - $mean)**2}; print $mean."\t".($sumdev/$nreads)**(1/2)."\n"}' & echo $sample; }` : this was the previous `[something2]`, in summary, it gets the average length and the standard deviation (sd) of all reads in a single `.fastq.gz` file. By steps:

  + `less $file` : read `.fastq.gz` file
  + `perl -ne 'print if /^[ATCG]+$/'` : print only the read sequences lines from the `.fastq.gz`
  + `awk '{ print length }'` : print (now only) the length of the reads before
  
  ```
  100
  178
  87
  ...
  ```
  
  + `{ [something3] & [something4]; }` : this structure combines the output of two commands, in this case, it combines the output of `[something3]` and `[something4]`
  +  `perl -ne 'BEGIN{$sum=0, $nreads=0, @dev=()} $nreads += 1; $sum += (split)[0]; push(@dev,(split)[0]); ` `END{ $sumdev=0; $mean = $sum/$nreads; foreach my $val (@dev) {$sumdev+= ($val - $mean)**2}; print $mean."\t".($sumdev/$nreads)**(1/2)."\n"}'` : corresponds to the `[something3]` before, it is a small perl code to sum all read lengths (var $sum) and store them (@dev) in order to compute the average length and the standard deviation of those
      + the `-n` switch is short for `while(read lines) {do something with each line}`, but `BEGIN{}` code stands before the while and `END{}` code after the while
 + `echo $sample` : corresponds to the `[something4]` before, it prints the name of the current sample being analyzed

* `perl -pe 's/(GSM.*)\n/$1\t/'` : combine the lines with the name of the sample and their avglen and sd (just changing a \n for a \t)

Until this point, the avglen and the sd of each .fastq.gz file where outputted, but there are (or could be) several .fastq.gz in each sample, e.g.:

```
HPIV3_infected_A549/GSM4462360	135.5	18.7216452268491
HPIV3_infected_A549/GSM4462360	130 19.7216452268491
HPIV3_infected_A549/GSM4462361	135.5	18.7216452268491
HPIV3_infected_A549/GSM4462362	122.666666666667	22.3308456325923
```

to get the mean avglen and the mean sd for all runs of a sample, I added the following:

* `sort` : sort previous output, according to `experiment/sample` name
* `perl -ne 'BEGIN{ $csample = "HPIV3_infected_A549/GSM4462360", $sumlen =0, $sumsd=0, $nfastqs=0 } if ((split)[0] eq $csample) { $nfastqs++; $sumlen += (split)[1]; $sumsd += (split)[2]; } else{ print $csample."\t".$sumlen/$nfastqs."\t".$sumsd/$nfastqs."\n"; $csample = (split)[0]; $nfastqs=1 ; $sumlen=(split)[1] ; $sumsd=(split)[2]} END{ print $csample."\t".$sumlen/$nfastqs."\t".$sumsd/$nfastqs."\n"}'` : this perl code computes the mean avglen and the mean sd of each line that belong to the same `experiment/sample`

final output example:

```
HPIV3_infected_A549/GSM4462360	117.1375	21.9739881152513
HPIV3_infected_A549/GSM4462361	135.5	18.7216452268491
HPIV3_infected_A549/GSM4462362	122.666666666667	22.3308456325923
```

\

***

This was run in each experiment folder, with sges named `reads_avgsdlen.sge` :

```{r eval=FALSE}
qsub reads_avgsdlen.sge
```

generating the following files:

```
IAV_infected_4_month_old_Ferret-reads_avgsdlen.tsv
ravgsd01_IAV_infected_4_month_old_Ferret.o169746
ravgsd01_IAV_infected_4_month_old_Ferret.po169746
```

\

***

Because the experiment dir name was missing, and to join all files with avglen and sd reads length info, I did:

```{r eval=FALSE}
for exp in *; do perl -pe 's/^./'$exp'/' $exp/$exp-reads_avgsdlen.tsv; done > samples_reads_avgsdlen.tsv
```

e.g :

```
./GSM4462360	117.1375	21.9739881152513 --> HPIV3_infected_A549/GSM4462360	117.1375	21.9739881152513
```

\

### **Kallisto**

[Kallisto web page](https://pachterlab.github.io/kallisto/about)

#### **Get Reference trasncriptome from ensembl**

Get [transcriptome reference data](https://m.ensembl.org/info/data/ftp/index.html) for human and ferret (Mustela putorius furo) (cdna) from the lastest version of Ensembl (v103 - Feb 2021):

cd = `/mnt/Citosina/amedina/mpadilla/`

```{r eval=FALSE}
mkdir resources/ensembl
mkdir resources/ensembl/cdna
cd resources/ensembl/cdna/
wget http://ftp.ensembl.org/pub/release-103/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz
wget http://ftp.ensembl.org/pub/release-103/fasta/mustela_putorius_furo/cdna/Mustela_putorius_furo.MusPutFur1.0.cdna.all.fa.gz
```

ref cdna data : `Homo_sapiens.GRCh38.cdna.all.fa.gz` and `Mustela_putorius_furo.MusPutFur1.0.cdna.all.fa.gz`

***

\

#### **Build index for human and ferret transcriptome**

cd = `/mnt/Citosina/amedina/mpadilla/resources`

```{r eval=FALSE}
mkdir kallisto
module load kallisto/0.45.0
kallisto index -i ./kallisto/index_kallisto45_ensemblv103_hsapiens_cdna ensembl/cdna103/Homo_sapiens.GRCh38.cdna.all.fa.gz
kallisto index -i ./kallisto/index_kallisto45_ensemblv103_mputoriusf_cdna ensembl/cdna103/Mustela_putorius_furo.MusPutFur1.0.cdna.all.fa.gz
```

* `index` builds an index from a FASTA formatted file of target sequences
* `-i` index name
* `ensembl/cdna/*.cdna.all.fa.gz` fasta file

***

\

#### **Count reads by sample**


Make directories for counts output:

cd = `/mnt/Citosina/amedina/mpadilla/COVID19/Data/counts`

```{r eval=FALSE}
perl -pe 'm/(.*)\t(.*)\t(.*)/; system("mkdir",$3) unless (-e $3); system("mkdir",$3."/".$2) unless (-e $3."/".$2)' ../metadata/inter-files/runs_samples_experiments.tsv
```

***

\

Generate file with pipelines to run `kallisto quant` by sample, from file done before with sample read length info, just adding to each line the rest of the command:

For Human samples:

```{r eval=FALSE}
grep -v "Ferret" ../trimmed2/samples_reads_avgsdlen.tsv | perl -pe 's/(.*)\t(.*)\t(.*)/kallisto quant -i \/mnt\/Citosina\/amedina\/mpadilla\/resources\/kallisto\/index_kallisto45_ensemblv102_hsapiens_cdna -o $1\/ --single -l $2 -s $3 \/mnt\/Citosina\/amedina\/mpadilla\/COVID19\/Data\/trimmed2\/$1\/*.fastq.gz/' > human_kallisto_quant.sge
```

For Ferret samples:

```{r eval=FALSE}
grep "Ferret" ../trimmed2/samples_reads_avgsdlen.tsv | perl -pe 's/(.*)\t(.*)\t(.*)/kallisto quant -i \/mnt\/Citosina\/amedina\/mpadilla\/resources\/kallisto\/index_kallisto45_ensemblv102_mputoriusf_cdna -o $1\/ --single -l $2 -s $3 \/mnt\/Citosina\/amedina\/mpadilla\/COVID19\/Data\/trimmed2\/$1\/*.fastq.gz/' > ferret_kallisto_quant.sge
```


***

\

General pipeline to run kallisto on samples:

```{r eval=FALSE}
kallisto quant -i /mnt/Citosina/amedina/mpadilla/resources/kallisto/index_kallisto45_ensemblv102_hsapiens_cdna -o /mnt/Citosina/amedina/mpadilla/COVID19/Data/counts/exp/sample/ --single -l [avglen] -s [meansd] *.fastq.gz
```

* `kallisto quant` runs the quantification algorithm
* `-i` Filename for the kallisto index to be used for quantification
* `-o` Directory to write output to
* `-b` Number of bootstrap samples (default: 0)
* `-l` Estimated average fragment length. I checked the fastqcs and it was around
* `-s` Estimated standard deviation of fragment length

```{r eval=FALSE}
```

***

\

Run jobs:

```{r eval=FALSE}
qsub human_kallisto_quant.sge
qsub ferret_kallisto_quant.sge
```

Output: `COVID19/Data/counts/experiment/sample/`:

```
abundance.h5  abundance.tsv  run_info.json
```

***

\



\

### **References**

[1] Blanco-Melo, D., Nilsson-Payant, B. E., Liu, W. C., Uhl, S., Hoagland, D., Møller, R., ... & Albrecht, R. A. (2020). Imbalanced host response to SARS-CoV-2 drives development of COVID-19. Cell, 181(5), 1036-1045.

[2] Williams, C. R., Baccarella, A., Parrish, J. Z., & Kim, C. C. (2016). Trimming of sequence reads alters RNA-Seq gene expression estimates. BMC bioinformatics, 17(1), 1-13.

